### PR TITLE
feat: Add String type

### PR DESCRIPTION
**Pull request by bbernays**

### CODE REVIEW

**Review by disq — APPROVED** [review]

File: src/scalar/text.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { Utf8 as ArrowString } from '@apache-arrow/esnext-esm';
+
+import { isInvalid, NULL_VALUE } from './util.js';
+
+export class Text {
```
Comment (erezrokah) [inline src/scalar/text.ts:5]: I added a new interface that should help implement scalars, can you try adding it, see:
https://github.com/cloudquery/plugin-sdk-javascript/blob/c7b301dfd20d4ce6dd1268090eedf0a4718fc2f9/src/scalar/bool.ts#L7

File: src/scalar/text.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  public get value(): string {
+    return this._value;
+  }
+
+  public set value(value: unknown) {
+    if (isInvalid(value)) {
+      this._valid = false;
+      return;
+    }
+    if (typeof value === 'string' || value instanceof String) {
```
Comment (erezrokah) [inline src/scalar/text.ts:31]: This means we only support settings strings to the string scalar.
We should support more types as in:

File: src/scalar/text.test.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { DataType } from '@apache-arrow/esnext-esm';
+import test from 'ava';
+
+import { Text } from './text.js';
+
+// eslint-disable-next-line unicorn/no-null
+[null, undefined].forEach((v) => {
+  test(`should set values to empty string when ${v} is passed`, (t) => {
+    const s = new Text(v);
+    t.is(s.value, '');
```
Comment (erezrokah) [inline src/scalar/text.test.ts:10]:
```suggestion
    t.is(s.value, '');
    t.is(s.valid, false);
```

File: src/scalar/text.test.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    t.true(DataType.isUtf8(s.dataType));
+  });
+});
+
+['', 'test string', String('asdf')].forEach((v, index) => {
```
Comment (erezrokah) [inline src/scalar/text.test.ts:15]:
```suggestion
['', 'test string', String('asdf'), new Test("test")].forEach((v, index) => {
```

File: src/scalar/text.ts (hunk 2 above)
Comment (erezrokah) [reply src/scalar/text.ts:31]: I think the equivalent to `Stringer` in Go is to check `Object.hasOwn(object1, 'toString')`, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn

**Review by erezrokah — CHANGES_REQUESTED** [review]
Thanks @bbernays, added a few comments

File: src/scalar/text.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    throw new Error(`Unable to set '${value}' as Text`);
+  }
+
+  public toString = (): string => {
+    if (this._valid) {
+      return this._value.toString();
+    }
+
+    return NULL_VALUE;
+  };
```
Comment (bbernays) [inline src/scalar/text.ts:57]: @erezrokah - Do you know why I had to change this method from:
```
  public toString() {
    if (this._valid) {
      return this._value.toString();
    }

    return NULL_VALUE;
  };
  ```
  
  In order for the `Object.hasOwn` to see that this class has a `toString` method?

File: src/scalar/text.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+  public set value(value: unknown) {
+    if (isInvalid(value)) {
+      this._valid = false;
+      return;
+    }
+
+    switch (typeof value) {
+      case 'object': {
+        if (value !== undefined && value !== null && Object.hasOwn(value, 'toString')) {
```
Comment (disq) [inline src/scalar/text.ts:35]:
```suggestion
        if (Object.hasOwn(value, 'toString')) {
```
We already check for invalid `value` above with `isInvalid()`

**Review by erezrokah — CHANGES_REQUESTED** [review]
Sorry, did a review yesterday but forgot to submit it 🤦

**Review by erezrokah — APPROVED** [review]

**Comment by erezrokah** [comment]
Going to merge this as I need this scalar for the sync

File: src/scalar/text.ts (hunk 6 above)
Comment (bbernays) [reply src/scalar/text.ts]: Without this check the typescript compiler/builder gave errors. If there is a way around that I would be happy to change it

**Pull request merged**